api/src/wfl/sink.clj

```diff
@@ -242,21 +243,105 @@
                                 :fromOutputs fromOutputs}))))
     (assoc request :dataset dataset')))
 
+(defn ^:private push-job [{:keys [details] :as _sink} job]
+  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+    (jdbc/insert! tx details job)))
+
+(defn ^:private peek-job-queue [{:keys [details] :as _sink}]
+  (let [query "SELECT * FROM %s
+               WHERE status = 'succeeded' AND consumed IS NULL
+               ORDER BY id LIMIT 1"]
+    (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+      (postgres/throw-unless-table-exists tx details)
+      (first (jdbc/query tx (format query details))))))
+
+(defn ^:private pop-job-queue!
+  [{:keys [details] :as _sink} {:keys [id] :as _job}]
+  {:pre [(some? id) (integer? id)]}
+  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+    (jdbc/update! tx details {:consumed (utc-now)} ["id = ?" id])))
+
+(defn ^:private update-datarepo-job-statuses
+  "Fetch and record the status of all 'running' jobs from tdr created by the
+   `sink`. Throws `UserException` when any new job status is `failed`."
+  [{:keys [details] :as _sink}]
+  (letfn [(read-active-jobs []
+            (let [query "SELECT id, job, workflow FROM %s
+                         WHERE status = 'running'
+                         ORDER BY id ASC"]
+              (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+                (map (juxt :id :job :workflow)
+                     (jdbc/query tx (format query details))))))
+          (write-job-status [job-id status]
+            (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+              (jdbc/update! tx details
+                            {:status status :updated (utc-now)}
+                            ["id = ?" job-id])))]
+    (doseq [[id job-id workflow] (read-active-jobs)]
+      (let [job    (datarepo/job-metadata job-id)
+            status (-> job :job_status str/lower-case)]
+        (write-job-status id status)
+        (when (= "failed" status)
+          (throw (UserException. "A DataRepo ingest job failed"
```

Comment:
This reminds me, maybe we should throw when TDR source snapshot creation jobs fail too (outside of the scope of this ticket).  More generally, a failure condition that would result in a stage's queue element never being eligible for consumption would benefit from an exception propagated.

Reply:
Yeah we probably should. Think right now we just log it.